### PR TITLE
AP_BattMonitor: UAVCAN support

### PR DESCRIPTION
**Pull request by EShamaev**
Support for voltage and current with CircuitStatus and BatteryInfo messages over UAVCAN.

### CODE REVIEW

**Comment by WickedShell** [comment]
I haven't had time to look at it in detail, but is there any reason why we couldn't just do a single UAVCAN backend that pulls either a battery status or a circuit status message? (with a preference for the battery one if it's available?)

**Comment by EShamaev** [comment]
The messages are very different and ID of circuit and Battery have absolutely different meanings.

**Comment by EShamaev** [comment]
But if there is a way to have 2 different IDs at the same time (one for circuit and one for battery) with current set of parameters, I will be happy to make such change.

**Comment by WickedShell** [comment]
I'm a bit confused, the same node wouldn't be sending both CircuitStatus and BatteryInfo would it? From an implementation perspective the only difference I see is that one of them happens to come with temperature information. AP_BattMonitor at the moment is explicitly and only intended for batteries (particularly meant for things that when they get low/consume a large amount of power we expect the vehicle to begin a failsafe action). If we want generic power logging around the system I have to wonder if that deserves a different library, (primarily so that we can have proper MAVLink reporting, and don't gain so many parameters per device we are monitoring).

**Comment by EShamaev** [comment]
I am referring not to node ID.
With CircuitStatus, same node can send many circuits ID (it is an electronics switch board with solid state circuit brakers and has 16 of circuits IDs) so this PR uses CircuitStatus ID and Battery ID and not the node ID, which seems much more appropriate to differentiate among those.

As for generic power logging, I am preparing such system, but it needs some more time.

**Comment by EShamaev** [comment]
For example "power bricks" used now has really nothing to do with batteries, but we treat voltage and current from those as a battery information.
The same is with CircuitStatus message. 
If it is reporting voltage and current from battery - we can well define it a battery.

**Comment by EShamaev** [comment]
> the same node wouldn't be sending both CircuitStatus and BatteryInfo would it?

That might be.
Smart battery that I use has generator input for charging and it is broadcasting it as a CircuitStatus together with BatteryInfo.

**Comment by WickedShell** [comment]
My hesitation is why we would accept a CircuitStatus as a battery input. We have the logical distinction, and I'm not so keen on blurring the line inside the BattMonitor library.

I'd argue the power bricks are a bit different, people tend to stick them upstream of all other components in the system, so it is actually monitoring the entire pack draw, which makes it a fair report on the battery status.

We only have support for 2 battery instances at the moment, and adding more eats parameters quickly (point of fact I don't think we have enough parameters to go past 5 in the current parameter group), which means that the 16 node CircutStatus wouldn't be fully monitored anyways...

**Comment by EShamaev** [comment]
2 instances are really enough, maybe only in extreme cases it can go over that.
Out of the 16 node CircutStatus I mentioned, only 2 are inputs from battery and the rest are outputs.
So as a use case, CircutStatus can represent the battery just like the powerbrick. And I actually expect that someone will release a "smart" power brick that will communicate over CAN. And in that case it will be CircutStatus message, not the Battery.
Anyway after combining the messages, the difference in AP_BattMonitor is really minimal and both can be supported without almost any overhead.

**Comment by liang-tang** [comment]
what is CircuitStatus?

**Comment by olliw42** [comment]
any chance this is being merged anytime soon?

I do have a working DIY uavcan power brick emitting power.CircuitStatus messages now, and would be eager to test it in real life
https://discuss.ardupilot.org/t/uavcan-for-hobbiests/16464/52

:)

**Comment by olliw42** [comment]
I'm not happy with the definition and usage of the uavcan.equipment.power messages

The usage of CircuitStatus I find illogical, and looks more like a misuse. The point IMHO is this: A "battery" has a specific additional function in AP (and likely any other system), in that it's status can trigger particular reactions. Specifically, e.g. a RTL. This would be so for any "battery", e.g. a smart battery, or a battery with a 3DR module to measure its voltage/current, or any other "thing" which powers the system and the voltage/current of which is used in that way. So, there is a clear cut conceptional difference between a "battery" and "any other electronics circuit". This is not reflected at all by the messages, and the above discussion and confusion is IMHO just a clear indication of that.

So, I would propose a set of "battery" messages, which is used for the purpose at hand, and to not use CircuitStatus. BatteryInfo is then for smart batteries (sad that it wasn't called SmartBatteryInfo). I suggest to add a new message "GenericBatteryInfo" (or named whatever), and to use that instead of CircuitStatus in this PR. For its fields I tentatively suggest

uint16 battery_id
float16 voltage
float16 current
float16 charge_consumed_mAh
uint8 status_flags

The addition of charge_consumed_mAh I think should be very helpful. Using current and the time difference between two messages does not allow a precise determination of the discharge. However, any node, which very likely is able to measure the current at much higher rate, can produce a much more precise measurement of the consumed charge. In my "old" days when I was flying a 450 heli a lot, which I equipped with my own-made telemetry, I found this to easily make a difference of 5-10%, which is a lot, IMHO.

This could also allow to not having to specify the "id", simplifying the setup enormously. E.g, if one can assume that only two batteries are in the system, one simply could catch all incoming messages of the battery type, and auto-fill in the SERIAL_NUM as it is done for a smart battery. If more than two batteries are connected , it's the very same situation as with all other "batteries".

If you think about it, the need to specify the correct circuit_id or battery_id makes the whole thing totally useless in practical terms. This is just another way to present the conceptual issue mentioned in the above.

Even though OT, I feel this needs to be added: It appears that essentially all messages of the "standard data types" which had been recently touched are quite ill designed, which I think should make one wonder if the adopted procedure which resulted in these standards is appropriate for defining a standard, and if they shouldn't be replaced by a more appropriate procedure. 
IMHO there should be some sort of "confirmed to work in practice" before a message is approved as standard.

**Comment by olliw42** [comment]
this is a bug in L389: https://github.com/ArduPilot/ardupilot/blob/4c3487cea190a62dd83656c2d8eee3c93718fad9/libraries/AP_UAVCAN/AP_UAVCAN.cpp#L389

it should be:  for (uint8_t i = 0; i < AP_UAVCAN_MAX_BI_NUMBER; i++) {

**Comment by EShamaev** [comment]
Removed CircuitStatus from battery monitoring.
New Electrical subsystem is coming soon.

**Comment by magicrub** [comment]
Hi @EShamaev can we get this rebased please?

**Comment by magicrub** [comment]
status update?

**Comment by magicrub** [comment]
@EShamaev ping. Needs rebase.

**Comment by EShamaev** [comment]
@magicrub Rebased, squashed.

File: libraries/AP_BattMonitor/AP_BattMonitor.h [hunk 1]
```diff
@@ -45,6 +47,23 @@ class AP_BattMonitor
         return *_singleton;
     }
 
+    // Battery monitor driver types
+    enum BattMonitor_Type {
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor.h]: Accidental regression? (This has moved to batt monitor params )

File: libraries/AP_BattMonitor/AP_BattMonitor_Params.cpp [hunk 2]
```diff
@@ -7,7 +7,8 @@ const AP_Param::GroupInfo AP_BattMonitor_Params::var_info[] = {
     // @Param: MONITOR
     // @DisplayName: Battery monitoring
     // @Description: Controls enabling monitoring of the battery's voltage and current
-    // @Values: 0:Disabled,3:Analog Voltage Only,4:Analog Voltage and Current,5:Solo,6:Bebop,7:SMBus-Maxell
+    // @Values: 0:Disabled,3:Analog Voltage Only,4:Analog Voltage and Current,5:Solo,6:Bebop,7:SMBus-Maxell,8:UAVCAN-BatteryInfo
+
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_Params.cpp]: Extra whitespace.

File: libraries/AP_BattMonitor/AP_BattMonitor_Params.cpp [hunk 3]
```diff
@@ -63,7 +64,8 @@ const AP_Param::GroupInfo AP_BattMonitor_Params::var_info[] = {
 
     // @Param: SERIAL_NUM
     // @DisplayName: Battery serial number
-    // @Description: Battery serial number, automatically filled in for SMBus batteries, otherwise will be -1
+    // @Description: Battery serial number, automatically filled in for SMBus batteries, otherwise will be -1. With UAVCAN it is the battery_id.
+
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_Params.cpp]: Extra whitespace.

File: libraries/AP_BattMonitor/AP_BattMonitor.h [hunk 4]
```diff
@@ -45,6 +47,23 @@ class AP_BattMonitor
...
+        BattMonitor_TYPE_ANALOG_VOLTAGE_ONLY        = 3,
+        BattMonitor_TYPE_ANALOG_VOLTAGE_AND_CURRENT = 4,
+        BattMonitor_TYPE_SOLO                       = 5,
+        BattMonitor_TYPE_BEBOP                      = 6,
+        BattMonitor_TYPE_MAXELL                     = 7,
+        BattMonitor_TYPE_UAVCAN_BatteryInfo         = 8
+    };
+
+    // low voltage sources (used for BATT_LOW_TYPE parameter)
+    enum BattMonitor_LowVoltage_Source {
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor.h]: Stale.

File: libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+    uint32_t tnow = AP_HAL::micros();
+    uint32_t dt = tnow - _state.last_time_micros;
+
+    // update total current drawn since startup
+    if (_state.last_time_micros != 0 && dt < 2000000) {
+        // .0002778 is 1/3600 (conversion to hours)
+        _state.current_total_mah += (float) ((double) _state.current_amps * (double) dt * (double) 0.0000002778);
+    }
+
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp]: You could log  from the `_state.consumed_wh = full_charge_capacity_wh - remaining_capacity_wh` (I get that this is new from when the PR first came in).

**Comment by EShamaev** [comment]
First, I tend to implement the whole interface so for example if there is addition - I also do removing. That helps to see if the overal idea of the implementation is working and done properly.
Second, they are logical and even if not used right now in platforms that we have - does not mean that they will not be used ever.
Example is: https://github.com/ArduPilot/ardupilot/pull/7768

However I understand your point of view, but would prefer them to stay.

**Comment by magicrub** [comment]
LGTM. Only weird part about it is BATT_SERIAL_NUM will require a reboot to take effect... only with UAVCAN. I'm not sure if that's worth putting in the description, or the @ REQUIRESREBOOT field, but should probably be noted in a comment somewhere

**Comment by magicrub** [comment]
merged!

**Comment by skyyuzhang** [comment]
I been researching AP_BattMonitor.UAVCAN.cpp for a while. I want to make uavcan batter monitor. I used libcanard to developer my project. data type which I used is "uavcan.equipment.power.BatteryInfo" I can get data volt,current,temperature in UAVCAN GUI correctly . but I can not get data in MissionPlanner.  it always shown "Bat 0.00v 0.0A 100%" also it is red word .  the bitrate is 1000000.  I check everything I can thought but It not work .can you help me ? thank you !!   the parameter I set are as follow

Copter-V3.6.2
CAN_D1_PROTOCOL =1
CAN_D2_PROTOCOL =1
CAN_D1_DRIVER =1
CAN_D2_DRIVER =2 
BATT_MONITOR =8
BATT_SERIAL_NUM=0      note: my uavcan battery id is 0(not node id!!!)

**Comment by EShamaev** [comment]
Please try battery id distinct from zero.

Current version is working for sure, so the problem seems to be in code other than AP.

**Comment by skyyuzhang** [comment]
> Please try battery id distinct from zero.
> 
> Current version is working for sure, so the problem seems to be in code other than AP.

Actually It was no use even I set battery distinct from zero . one of thing is I can get data on UAVCAN GUI Tool but I can not get data on mission planner. I have been finding the problem but no use . does libcanard supported ardupilot? or others  problem? thank god ! I can not get this done . please help~~~~ thank you!!!

**Comment by 475241512** [comment]
> > Please try battery id distinct from zero.
> > Current version is working for sure, so the problem seems to be in code other than AP.
> 
> Actually It was no use even I set battery distinct from zero . one of thing is I can get data on UAVCAN GUI Tool but I can not get data on mission planner. I have been finding the problem but no use . does libcanard supported ardupilot? or others problem? thank god ! I can not get this done . please help~~~~ thank you!!!

Have you solved the problem？I used libcanard "uavcan.equipment.power.BatteryInfo"datatype,It can be get and view in UAVCAN GUI,but mission planner always shown "0.00V,0.0A,100%",What did I do wrong? Do I have any details to pay attention to?thanks.

**Pull request merged**